### PR TITLE
UFAL/Autocomplete enhancement

### PR DESCRIPTION
**Pull request by milanmajchrak**
| Phases            | MP | MM  |  MB  | MR |   JM | Total  |
|-----------------|----:|----:|----:|-----:|-----:|-------:|
| ETA                  |  0  |  0  |    0 |     0 |      0 |        0 |
| Developing      |  0  |  0  |    0 |    0 |      0 |         0 |
| Review             |  0  |  0  |    0 |    0 |      0 |         0 |
| Total                |   -  |   -  |   -   |  -    |   -    |         0 |
| ETA est.             |      |      |       |       |         |         0 |
| ETA cust.           |   -  |   -  |   -  |   -   |   -     |        0 |

### CODE REVIEW

File: src/app/core/data/metadata-value-data.service.ts [hunk 1]
```diff
@@ -59,22 +59,22 @@ export class MetadataValueDataService extends BaseDataService<MetadataValue> imp
...
-      if (!isNotEmpty(metadataFields[0]) && !isNotEmpty(metadataFields[1])) {
+      if (!isNotEmpty(metadataFields?.[0]) && !isNotEmpty(metadataFields?.[1])) {
         return EMPTY;
       }
 
       // add value to the request params
-      schemaRP.fieldValue = metadataFields[0];
-      elementRP.fieldValue = metadataFields[1];
-      qualifierRP.fieldValue = isNotEmpty(metadataFields[2]) ? metadataFields[2] : null;
+      schemaRP.fieldValue = metadataFields?.[0];
```
Comment (vidiecan) [inline src/app/core/data/metadata-value-data.service.ts:75]: required?

File: src/app/core/data/metadata-value-data.service.ts [hunk 2]
```diff
@@ -59,22 +59,22 @@ export class MetadataValueDataService extends BaseDataService<MetadataValue> imp
...
+      const metadataFields = metadataName?.split('.');
 
       const schemaRP = new RequestParam('schema', '');
       const elementRP = new RequestParam('element', '');
       const qualifierRP = new RequestParam('qualifier', '');
       const termRP = new RequestParam('searchValue', term);
 
       // schema and element are mandatory - cannot be empty
-      if (!isNotEmpty(metadataFields[0]) && !isNotEmpty(metadataFields[1])) {
+      if (!isNotEmpty(metadataFields?.[0]) && !isNotEmpty(metadataFields?.[1])) {
```
Comment (vidiecan) [inline src/app/core/data/metadata-value-data.service.ts]: the more negation the better?
! not
!! not not ;)

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/sponsor-autocomplete/ds-dynamic-sponsor-autocomplete.component.ts [hunk 3]
```diff
@@ -37,18 +41,22 @@ export class DsDynamicSponsorAutocompleteComponent extends DsDynamicAutocomplete
...
+      lookupRelationService, requestService, rdbService, halService, configurationService, translateService);
   }
 
   /**
    * From suggestion update model: 1. openAIRE -> compose input from suggestion value,
    *                               2. metadata suggestion -> update as suggestion value.
    * @param updateValue
    */
-  updateModel(updateValue) {
+  async updateModel(updateValue) {
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/sponsor-autocomplete/ds-dynamic-sponsor-autocomplete.component.ts]: why

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/ds-dynamic-autocomplete.component.ts [hunk 4]
```diff
@@ -129,6 +190,7 @@ export class DsDynamicAutocompleteComponent extends DsDynamicTagComponent implem
    * @param x
    */
   formatter = (x: { display: string }) => {
+    console.log('formatter', x);
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/ds-dynamic-autocomplete.component.ts]: ?

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: src/app/core/data/metadata-value-data.service.ts (hunk 1 above)
Comment (milanmajchrak) [reply src/app/core/data/metadata-value-data.service.ts:75]: I do not understand your question

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/sponsor-autocomplete/ds-dynamic-sponsor-autocomplete.component.ts (hunk 3 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/sponsor-autocomplete/ds-dynamic-sponsor-autocomplete.component.ts]: It's mistake, thank you.

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/ds-dynamic-autocomplete.component.ts (hunk 4 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/ds-dynamic-autocomplete.component.ts]: It's mistake, thank you.

File: src/app/core/data/metadata-value-data.service.ts (hunk 2 above)
Comment (milanmajchrak) [reply src/app/core/data/metadata-value-data.service.ts]: Fixed, thank you

File: src/app/core/data/metadata-value-data.service.ts (hunk 1 above)
Comment (vidiecan) [reply src/app/core/data/metadata-value-data.service.ts:75]: is `?` required here

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/ds-dynamic-autocomplete.component.ts [hunk 5]
```diff
@@ -1,18 +1,53 @@
...
+import { CANONICAL_PREFIX_KEY } from '../../../../../handle.service';
+import { ConfigurationProperty } from '../../../../../../core/shared/configuration-property.model';
+import { DsDynamicAutocompleteService } from './ds-dynamic-autocomplete.service';
+import { TranslateService } from '@ngx-translate/core';
+
+/**
+ * Prefix for custom autocomplete definition from the `submission-forms.xml`.
+ * <input-type autocomplete-custom="solr-handle_title_ac">autocomplete</input-type>
+ */
+const AUTOCOMPLETE_CUSTOM_HANDLE_TITLE = 'solr-handle_title_ac';
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/ds-dynamic-autocomplete.component.ts:39]: @milanmajchrak TT about how it works, can we specify field we want autocomplate from from UI? unsafe

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/ds-dynamic-autocomplete.component.ts (hunk 5 above)
Comment (milanmajchrak) [reply src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/ds-dynamic-autocomplete.component.ts:39]: `AUTOCOMPLETE_CUSTOM_HANDLE_TITLE` is used only for prettifying the suggestion.

File: src/app/core/data/metadata-value-data.service.ts (hunk 1 above)
Comment (milanmajchrak) [reply src/app/core/data/metadata-value-data.service.ts:75]: During development it throws an error, so not in production, but I'd like to keep it there

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/sponsor-autocomplete/ds-dynamic-sponsor-autocomplete.component.ts [hunk 6]
```diff
@@ -77,7 +85,7 @@ export class DsDynamicSponsorAutocompleteComponent extends DsDynamicAutocomplete
       fundingProjectCode = fundingFields?.[1];
       fundingName = fundingFields?.[3];
     }
-    return DsDynamicAutocompleteService.pretifySuggestion(fundingProjectCode, fundingName, this.translateService);
+    return DsDynamicAutocompleteService.pretifySponsorSuggestion(fundingProjectCode, fundingName, this.translateService);
```
Comment (vidiecan) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/sponsor-autocomplete/ds-dynamic-sponsor-autocomplete.component.ts]: Prettify

**Review by vidiecan — CHANGES_REQUESTED** [review]

**Review by vidiecan — APPROVED** [review]

**Pull request merged**